File: backend/src/graduate_entrance/schemas/planning.py
from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

MaterialType = Literal[
    "textbook",
    "exercise_book",
    "past_paper",
    "course",
    "vocabulary",
    "other",
]
TaskType = Literal[
    "reading",
    "practice",
    "dictation",
    "past_paper",
    "memorization",
    "review",
]


class PlanningSubjectRead(BaseModel):
    id: UUID
    code: str
    name: str
    order: int


class PhaseSubjectRatioInput(BaseModel):
    subject_id: UUID
    percentage: int = Field(ge=0, le=100)


class PlanPhaseInput(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    start_date: date
    end_date: date
    description: str = Field(default="", max_length=4000)
    milestones: list[str] = Field(default_factory=list, max_length=20)
    allow_new_tasks: bool = True
    order: int = Field(default=0, ge=0)
    subject_ratios: list[PhaseSubjectRatioInput] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_phase(self) -> "PlanPhaseInput":
        if self.start_date > self.end_date:
            raise ValueError("start_date must not be after end_date")
        subject_ids = [ratio.subject_id for ratio in self.subject_ratios]
        if len(subject_ids) != len(set(subject_ids)):
            raise ValueError("subject ratios must contain unique subjects")
        total = sum(ratio.percentage for ratio in self.subject_ratios)
        if self.allow_new_tasks and total != 100:
            raise ValueError("subject ratios must total 100 when new tasks are allowed")
        if not self.allow_new_tasks and total not in (0, 100):
            raise ValueError("subject ratios must be empty or total 100")
        return self


class PlanPhaseRead(PlanPhaseInput):
    id: UUID


class AvailabilityRuleInput(BaseModel):
    weekday: int = Field(ge=0, le=6)
    available_minutes: int = Field(ge=0, le=1440)


class AvailabilityPeriodInput(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    start_date: date
    end_date: date
    weekly_target_minutes: int = Field(ge=0, le=10080)
    order: int = Field(default=0, ge=0)
    rules: list[AvailabilityRuleInput]

    @model_validator(mode="after")
    def validate_period(self) -> "AvailabilityPeriodInput":
        if self.start_date > self.end_date:
            raise ValueError("start_date must not be after end_date")
        weekdays = [rule.weekday for rule in self.rules]
        if sorted(weekdays) != list(range(7)):
            raise ValueError("availability rules must contain weekdays 0 through 6 exactly once")
        return self


class AvailabilityPeriodRead(AvailabilityPeriodInput):
    id: UUID


class AvailabilityExceptionInput(BaseModel):
    date: date
    available_minutes: int = Field(ge=0, le=1440)
    reason: str = Field(default="", max_length=240)


class AvailabilityExceptionRead(AvailabilityExceptionInput):
    id: UUID


class MaterialInput(BaseModel):
    subject_id: UUID | None = None
    name: str = Field(min_length=1, max_length=120)
    material_type: MaterialType
    source: str = Field(default="", max_length=240)
    description: str = Field(default="", max_length=4000)
    active: bool = True
    order: int = Field(default=0, ge=0)


class MaterialRead(MaterialInput):
    id: UUID


class TaskTemplateInput(BaseModel):
    subject_id: UUID
    material_id: UUID | None = None
    name: str = Field(min_length=1, max_length=120)
    task_type: TaskType
    default_est_minutes: int = Field(gt=0, le=1440)
    description: str = Field(default="", max_length=4000)
    active: bool = True
    order: int = Field(default=0, ge=0)
    phase_ids: list[UUID] = Field(min_length=1)

    @model_validator(mode="after")
    def validate_phase_ids(self) -> "TaskTemplateInput":
        if len(self.phase_ids) != len(set(self.phase_ids)):
            raise ValueError("phase_ids must be unique")
        return self


class TaskTemplateRead(TaskTemplateInput):
    id: UUID


class PlanningConfigResponse(BaseModel):
    subjects: list[PlanningSubjectRead]
    phases: list[PlanPhaseRead]
    availability_periods: list[AvailabilityPeriodRead]
    availability_exceptions: list[AvailabilityExceptionRead]
    materials: list[MaterialRead]
    task_templates: list[TaskTemplateRead]

File: web/src/services/planning.ts
import { apiFetch } from './api'

export interface PlanningSubject {
  id: string
  code: string
  name: string
  order: number
}

export interface PhaseSubjectRatio {
  subject_id: string
  percentage: number
}

export interface PlanPhaseInput {
  name: string
  start_date: string
  end_date: string
  description: string
  milestones: string[]
  allow_new_tasks: boolean
  order: number
  subject_ratios: PhaseSubjectRatio[]
}

export interface PlanPhase extends PlanPhaseInput {
  id: string
}

export interface AvailabilityRule {
  weekday: number
  available_minutes: number
}

export interface AvailabilityPeriodInput {
  name: string
  start_date: string
  end_date: string
  weekly_target_minutes: number
  order: number
  rules: AvailabilityRule[]
}

export interface AvailabilityPeriod extends AvailabilityPeriodInput {
  id: string
}

export interface AvailabilityExceptionInput {
  date: string
  available_minutes: number
  reason: string
}

export interface AvailabilityException extends AvailabilityExceptionInput {
  id: string
}

export type MaterialType =
  | 'textbook'
  | 'exercise_book'
  | 'past_paper'
  | 'course'
  | 'vocabulary'
  | 'other'

export interface MaterialInput {
  subject_id: string | null
  module_id?: string | null
  name: string
  material_type: MaterialType
  source: string
  description: string
  active: boolean
  order: number
}

export interface Material extends MaterialInput {
  id: string
}

export type TaskType =
  | 'reading'
  | 'practice'
  | 'dictation'
  | 'past_paper'
  | 'memorization'
  | 'review'

export interface TaskTemplateInput {
  subject_id: string
  material_id: string | null
  name: string
  task_type: TaskType
  default_est_minutes: number
  description: string
  active: boolean
  order: number
  phase_ids: string[]
}

export interface TaskTemplate extends TaskTemplateInput {
  id: string
}

export interface PlanningConfig {
  subjects: PlanningSubject[]
  phases: PlanPhase[]
  availability_periods: AvailabilityPeriod[]
  availability_exceptions: AvailabilityException[]
  materials: Material[]
  task_templates: TaskTemplate[]
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(`/api/planning${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  })
  if (!response.ok) {
    throw new Error(`Planning request failed with status ${response.status}`)
  }
  if (response.status === 204) {
    return undefined as T
  }
  return (await response.json()) as T
}

export function fetchPlanningConfig(): Promise<PlanningConfig> {
  return request<PlanningConfig>('/config')
}

export function createPlanPhase(payload: PlanPhaseInput): Promise<PlanPhase> {
  return request<PlanPhase>('/phases', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function deletePlanPhase(id: string): Promise<void> {
  return request<void>(`/phases/${id}`, { method: 'DELETE' })
}

export function createAvailabilityPeriod(
  payload: AvailabilityPeriodInput,
): Promise<AvailabilityPeriod> {
  return request<AvailabilityPeriod>('/availability-periods', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function deleteAvailabilityPeriod(id: string): Promise<void> {
  return request<void>(`/availability-periods/${id}`, { method: 'DELETE' })
}

export function createAvailabilityException(
  payload: AvailabilityExceptionInput,
): Promise<AvailabilityException> {
  return request<AvailabilityException>('/availability-exceptions', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function deleteAvailabilityException(id: string): Promise<void> {
  return request<void>(`/availability-exceptions/${id}`, { method: 'DELETE' })
}

export function createMaterial(payload: MaterialInput): Promise<Material> {
  return request<Material>('/materials', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function deleteMaterial(id: string): Promise<void> {
  return request<void>(`/materials/${id}`, { method: 'DELETE' })
}

export function createTaskTemplate(payload: TaskTemplateInput): Promise<TaskTemplate> {
  return request<TaskTemplate>('/task-templates', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function deleteTaskTemplate(id: string): Promise<void> {
  return request<void>(`/task-templates/${id}`, { method: 'DELETE' })
}
